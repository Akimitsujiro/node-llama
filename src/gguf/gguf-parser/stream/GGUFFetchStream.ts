import retry from "async-retry";
import AwaitLock from "await-lock";
import GgufBaseStream, {ALLOCATION_SIZE} from "./GGUFBaseStream.js";

type GGUFFetchStreamOptions = {
    retry: retry.Options
};

export default class GGUFFetchStream extends GgufBaseStream {
    private _lock = new ("default" in AwaitLock && AwaitLock || AwaitLock)();

    constructor(public readonly url: string, public readonly options: Partial<GGUFFetchStreamOptions> = {}) {
        super();
    }

    override async readNBytes(numBytes: number, offset = 0): Promise<Buffer> {
        await this._lock.acquireAsync();

        try {
            if (offset + numBytes < this._buffer.length) {
                return this._buffer.subarray(offset, offset + numBytes);
            }

            const fetchMissingBytes = await retry(async () => {
                return await this._fetchBytesWithoutRetry(this._buffer.length, offset + numBytes + ALLOCATION_SIZE);
            }, this.options.retry);

            this._addToBuffer(fetchMissingBytes);
            return this.readNBytes(numBytes, offset);
        } finally {
            this._lock.release();
        }
    }

    private async _fetchBytesWithoutRetry(start: number, end: number) {
        const response = await fetch(this.url, {
            headers: {
                Range: `bytes=${start}-${end}`,
                accept: "*/*"
            }
        });
        const arrayBuffer = await response.arrayBuffer();
        return Buffer.from(arrayBuffer);
    }
}


