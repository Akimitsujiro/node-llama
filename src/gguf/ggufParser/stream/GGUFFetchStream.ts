import retry from "async-retry";
import {withLock} from "lifecycle-utils";

import GgufBaseStream, {ALLOCATION_SIZE} from "./GGUFBaseStream.js";

type GGUFFetchStreamOptions = {
    retry: retry.Options
};

export default class GGUFFetchStream extends GgufBaseStream {
    public readonly url: string;
    public readonly options: Partial<GGUFFetchStreamOptions> = {};

    public constructor(url: string, options: Partial<GGUFFetchStreamOptions> = {}) {
        super();
        this.options = options;
        this.url = url;
    }

    public override async readNBytes(numBytes: number, offset = 0): Promise<Buffer> {
        return await withLock(this, "_lock", async function readNBytesWithoutLock(): Promise<Buffer> {
            if (offset + numBytes < this._buffer.length) {
                return this._buffer.subarray(offset, offset + numBytes);
            }

            const fetchMissingBytes = await retry(async () => {
                return await this._fetchBytesWithoutRetry(this._buffer.length, offset + numBytes + ALLOCATION_SIZE);
            }, this.options.retry);

            this._addToBuffer(fetchMissingBytes);
            return await readNBytesWithoutLock.call(this);
        });
    }

    private async _fetchBytesWithoutRetry(start: number, end: number) {
        const response = await fetch(this.url, {
            headers: {
                Range: `bytes=${start}-${end}`,
                accept: "*/*"
            }
        });
        const arrayBuffer = await response.arrayBuffer();
        return Buffer.from(arrayBuffer);
    }
}


