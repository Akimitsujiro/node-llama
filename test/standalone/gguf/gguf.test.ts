import {describe, expect, it, test} from "vitest";
import GGUFParser from "../../../src/gguf/ggufParser/GGUFParser.js";
import GGUFFetchStream from "../../../src/gguf/ggufParser/stream/GGUFFetchStream.js";

const remoteGGUFModel = "https://huggingface.co/TheBloke/Falcon-180B-Chat-GGUF/resolve/main/falcon-180b-chat.Q6_K.gguf-split-a?download=true";

describe("GGUF Parser", async () => {
    test("Magic should be GGUF remote model", {timeout: 1000 * 60 * 10}, async () => {
        const stream = new GGUFFetchStream(remoteGGUFModel);

        const magic = await stream.readNBytes(4);
        const magicText = String.fromCharCode(...magic);

        expect(magicText).toBe("GGUF");
    });

    it("should parse remote gguf model", async () => {
        const stream = new GGUFFetchStream(remoteGGUFModel);

        const ggufParser = new GGUFParser(stream);
        const metadata = await ggufParser.parseMetadata();

        expect(metadata).toMatchSnapshot();
    });
});
