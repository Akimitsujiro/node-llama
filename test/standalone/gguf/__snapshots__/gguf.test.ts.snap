// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GGUF Parser > should parse remote gguf model 1`] = `
{
  "metadata": {
    "falcon": {
      "attention": {
        "head_count": 232,
        "head_count_kv": 8,
        "layer_norm_epsilon": 0.000009999999747378752,
      },
      "block_count": 80,
      "context_length": 2048,
      "embedding_length": 14848,
      "feed_forward_length": 59392,
      "tensor_data_layout": "jploski",
    },
    "general": {
      "architecture": "falcon",
      "file_type": "MOSTLY_Q6_K",
      "name": "Falcon",
      "quantization_version": 2,
    },
    "tensorCount": 644,
    "tokenizer": {
      "ggml": {
        "eos_token_id": 11,
        "model": "gpt2",
      },
    },
    "version": 2,
  },
  "metadataSize": 2547826,
}
`;
